tools/clang/lib/AST/ASTDumper.cpp

```diff
@@ -1072,12 +1078,10 @@ void ASTDumper::dumpHLSLUnusualAnnotations(const ArrayRef<hlsl::UnusualAnnotatio
 
 void ASTDumper::dumpDecl(const Decl *D) {
   // HLSL Change Starts: Don't display decls with invalid SourceLocations.
-  if (D && D->getDeclContext() &&
+  if (!DumpImplicitTopLevelDecls && D->getDeclContext() &&
```

Comment:
Can we move the top-level implicit decl skip up to `dumpDeclContext` instead?  Then I think it simplifies things, doesn't it?  We no longer have to default `DumpImplicitTopLevelDecl` to `true` in order to make `dumpDecl` work under the debugger, for instance.